pkg/controllers/management/aks/aks_cluster_handler_test.go

```diff
@@ -0,0 +1,259 @@
+ */
+func Test_onClusterChange_ClusterIsNil(t *testing.T) {
+	cluster, _ := mockOperatorController.onClusterChange("", nil)
+	if cluster != nil {
+		t.Error("cluster should have returned nil")
+	}
+}
+
+func Test_onClusterChange_AKSConfigIsNil(t *testing.T) {
+	mockCluster := &v3.Cluster{
+		Spec: v3.ClusterSpec{
+			AKSConfig: nil,
+		},
+	}
+
+	cluster, _ := mockOperatorController.onClusterChange("", mockCluster)
+	if cluster != mockCluster {
+		t.Error("cluster should have returned with no update")
+	}
+}
+
+func Test_onClusterChange_Default(t *testing.T) {
+
+	// setup
+	// create an instance of the operator controller with mock data to simulate the onChangeCluster function reacting
+	// to a real cluster!
+	mockOperatorController = getMockAksOperatorController("default")
+
+	mockCluster, err := getMockV3Cluster(MockDefaultClusterFilename)
+	if err != nil {
+		t.Error("error getting mock v3 cluster: ", err.Error())
+	}
+
+	// run test
+	cluster, err := mockOperatorController.onClusterChange("", &mockCluster)
+
+	// validate results
+	if cluster.Status.Conditions[0].Status != "Unknown" || err != nil {
+		t.Error("status should be Unknown and cluster returned successfully: " + err.Error())
+	}
+}
+
+func Test_onClusterChange_Create(t *testing.T) {
+	mockOperatorController = getMockAksOperatorController("create")
+	mockCluster, err := getMockV3Cluster(MockCreateClusterFilename)
+	if err != nil {
+		t.Error("error getting mock v3 cluster: ", err.Error())
+	}
+
+	cluster, err := mockOperatorController.onClusterChange("", &mockCluster)
+
+	if cluster.Status.Conditions[0].Status != "Unknown" || err != nil {
+		t.Error("provisioned status should be Unknown and cluster returned successfully: " + err.Error())
+	}
+}
+
+func Test_onClusterChange_Active(t *testing.T) {
+	mockOperatorController = getMockAksOperatorController("active")
+	mockCluster, err := getMockV3Cluster(MockActiveClusterFilename)
+	if err != nil {
+		t.Error("error getting mock v3 cluster: ", err.Error())
+	}
+
+	cluster, err := mockOperatorController.onClusterChange("", &mockCluster)
+
+	if cluster.Status.Conditions[0].Status != "True" ||
+		cluster.Status.Conditions[1].Status != "True" || err != nil {
+		t.Error("provisioned and updated status should be True and cluster returned successfully: " + err.Error())
+	}
+}
+
+func Test_onClusterChange_UpdateNodePool(t *testing.T) {
+	mockOperatorController = getMockAksOperatorController("update")
+	mockCluster, err := getMockV3Cluster(MockUpdateClusterFilename)
+	if err != nil {
+		t.Error("error getting mock v3 cluster: ", err.Error())
+	}
+
+	cluster, err := mockOperatorController.onClusterChange("", &mockCluster)
+
+	// check that cluster is updating node pool
+	if cluster.Status.Conditions[0].Status != "True" ||
+		cluster.Status.Conditions[1].Status != "Unknown" || err != nil {
+		t.Error("provisioned status should be True, updated status should be Unknown and cluster returned successfully: " + err.Error())
+	}
+}
```

Comment:
```suggestion
	if err != nil {
		t.Errorf("error running onClusterChange: %s", err)
	}
	if !rke2.Provisioned.IsTrue(cluster) || !rke2.Updated.IsUnknown(cluster) {
		t.Errorf("provisioned status should be True, updated status should be Unknown and cluster returned successfully: %s", err)
	}
}
```
I would break up these errors, since an error from `onClusterChange` should not be related to the conditions. I would also use the more idiomatic approach to the conditions (as these tests grow and expand, manually managing condition indices will become more cumbersome). Also using `Errorf("... %s", err)` instead is merely a stylistic choice, but it's one I prefer: it's more concise, usually allows for better git diffs, and the intention is clearer.

Reply:
Sometimes `config, err` returned from a handler means an API call failed yes but it could also mean the cluster isn't in the state it should be so the cluster is returned to wait for updates.

Example:
```
cluster, err := e.CheckCrdReady(cluster, "aks")
	if err != nil {
		return cluster, err
	}
```

But in general yes, I think this is a good idea.